src/components/accordion/accordion.component.js

```diff
@@ -0,0 +1,146 @@
+import React, {
+  useState, useRef, useEffect, useCallback
+} from 'react';
+import PropTypes from 'prop-types';
+
+import OptionsHelper from '../../utils/helpers/options-helper';
+import createGuid from '../../utils/helpers/guid';
+import Events from '../../utils/helpers/events';
+import {
+  StyledAccordionContainer,
+  StyledAccordionTitleContainer,
+  StyledAccordionTitle,
+  StyledAccordionIcon,
+  StyledAccordionContentContainer,
+  StyledAccordionContent
+} from './accordion.style';
+
+const Accordion = ({
+  defaultExpanded,
+  expanded,
+  onChange,
+  children,
+  id,
+  iconType,
+  iconAlign,
+  styleOverride,
+  type,
+  title
+}) => {
+  const isControlled = expanded !== undefined;
+
+  const [isExpandedInternal, setIsExpandedInternal] = useState(defaultExpanded || false);
+
+  const [contentHeight, setContentHeight] = useState(0);
+
+  const accordionContent = useRef(null);
+
+  useEffect(() => {
+    setContentHeight(!expanded ? 0 : accordionContent.current.scrollHeight);
+  }, [expanded]);
+
+  const isExpanded = isControlled ? expanded : isExpandedInternal;
+
+  const toggleAccordion = useCallback((ev) => {
+    if (!isControlled) setIsExpandedInternal(!isExpanded);
+    if (onChange) onChange(ev, !isExpanded);
+
+    setContentHeight(isExpanded ? 0 : accordionContent.current.scrollHeight);
+  }, [isControlled, isExpanded, onChange]);
+
+  const handleKeyDown = useCallback((ev) => {
+    if (Events.isEnterKey(ev)) {
+      toggleAccordion(ev);
+    }
+  }, [toggleAccordion]);
+
+  const guid = useRef(createGuid());
```

Comment:
@mkrds does the `guid` need to be a ref if it's not being used inside a hook?

Reply:
I think it's better if it's not generated on each render as there's no need for it. Also there's a accessibility connection between clickable header and expandable content defined by by `aria-controls={ contentId }` -> `id={ contentId }` and `id={ headerId }` -> `aria-labelledby={ headerId }`. I guess it might mess up with screen readers if the id's are regenerated on each expand/collapse action.

Reply:
Ah I get you. I think the same can be achieved with ` const [guid] = useState(createGuid());` but it doesn't matter, was just seeing if I was missing something, thanks 😄

Reply:
Both of these methods call `createGuid()` on each render. The only solution that I know of is to put it a
```js
useEffect(() => {
  setGuid(createGuid());
},[]);
```

Reply:
Even though `const guid = useRef(createGuid());` calls `createGuid()` on each render the guid is assigned to the ref on the first call so the only drawback of this solution is performance impact which is negligible as `createGuid()` takes only about 0.01ms

```js
useEffect(() => {
  setGuid(createGuid());
},[]);
```
Solves the problem with calling `createGuid()` multiple times but on the first render the id is `undefined`

Reply:
`useRef` it is then, let's optimise if it starts causing an issue.